### ISSUE TITLE
Strange behaviors (using_moving_average and last_gamma)

### ISSUE THREAD

**Issue by Cadene** [switchablenorms/Switchable-Normalization#3]
Hello,

I am trying to use SwitchNorm instead of BatchNorm in my current project with pytorch v0.4.0.
I encounter two strange behaviors and I would like to know if I am the only one before starting a more in-depth debugging.

The first behavior is using `using_moving_average=True` and `last_gamma=True`. My training and testing losses stay high (both around 14).

The second behavior is using `using_moving_average=False` and `last_gamma=False`. My training loss decreases in a normal way (around 0), but my testing loss stays really high (around 24)
I found a hack to be able to use the layer: I modified it so that it computes the batch statistics in eval mode as it is done in [nn.BatchNorm2d](https://pytorch.org/docs/stable/nn.html#torch.nn.BatchNorm2d).

Thanks for your help

**Comment by JiaminRen** [comment]
`using_moving_average=True`: during training SN layer keeps running estimates of its computed mean and variance, which are then used for normalization during evaluation.
`using_moving_average=False` : it computes the true mean and variance on a sufficiently large training batch, which are then used for evaluation. you can use `sn_helper`  in `main.py` after training.
`last_gamma`: whether to initialize the gamma of the last SN layer in each bottleneck to zero.
so, you can set `using_moving_average=True` and `last_gamma=False` for general use.

**Comment by Cadene** [comment]
Oh ok, so the general use is `using_moving_average=True` and `last_gamma=False` (SwitchNorm seems to work significantly better than BatchNorm on my task for now 💃).
And also if I want to try `using_moving_average=False`, I need to use the [sn_helper](https://github.com/switchablenorms/Switchable-Normalization/blob/master/main.py#L235) function before each evaluation. All right!

Last thing, what is the insight behind using `last_gamma=True` to initialize the gamma of the last SN layer in each bottleneck to zero?



Thanks again for your help.

**Comment by JiaminRen** [comment]
Setting `gamma=0`  in the last Normalization layer of each residual block causes the forward/backward signal initially to propagate through the identity shortcut of ResNets, which can ease optimization at the start of training. This insight is from [the paper](https://arxiv.org/pdf/1706.02677.pdf).

**Comment by Banyueqin** [comment]
@JiaminRen 
sn_helper in main.py after training
where is the main.py?

**Comment by seraphzl** [comment]
@Cadene @JiaminRen 
Where is the main.py?
I would use the sn_helper function.

**Comment by JiaminRen** [comment]
> @Cadene @JiaminRen
> Where is the main.py?
> I would use the sn_helper function.

you can use [the previous branch] (https://github.com/switchablenorms/Switchable-Normalization/tree/cd7115913808c5f501f2f37bcd22641e4131060a)

**Comment by JiaminRen** [comment]
you can use the previous commit (https://github.com/switchablenorms/Switchable-Normalization/tree/cd7115913808c5f501f2f37bcd22641e4131060a).
